crates/sp-lightclient/src/lib.rs

```diff
@@ -0,0 +1,299 @@
+
+//! Light client substrate primitives for Subspace.
+#![forbid(unsafe_code)]
+#![warn(rust_2018_idioms, missing_docs)]
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::{Decode, Encode};
+use scale_info::TypeInfo;
+use sp_consensus_subspace::digests::{
+    extract_pre_digest, extract_subspace_digest_items, CompatibleDigestItem, Error as DigestError,
+    PreDigest, SubspaceDigestItems,
+};
+use sp_consensus_subspace::{FarmerPublicKey, FarmerSignature};
+use sp_runtime::traits::Header as HeaderT;
+use sp_std::cmp::Ordering;
+use subspace_core_primitives::{PublicKey, Randomness, RewardSignature, Salt};
+use subspace_solving::{derive_global_challenge, derive_target, REWARD_SIGNING_CONTEXT};
+use subspace_verification::{check_reward_signature, verify_solution, VerifySolutionParams};
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(test)]
+mod mock;
+
+// TODO(ved): move them to consensus primitives and change usages across
+/// Type of solution range
+type SolutionRange = u64;
+
+/// The size of data in one piece (in bytes).
+type RecordSize = u32;
+
+/// The size of encoded and plotted piece in segments of this size (in bytes).
+type SegmentSize = u32;
+
+/// BlockWeight type for fork choice rules
+type BlockWeight = u128;
+
+/// HeaderExt describes an extended block chain header at a specific height along with some computed values.
+#[derive(Default, Debug, Encode, Decode, Clone, Eq, PartialEq, TypeInfo)]
+pub struct HeaderExt<Header> {
+    /// Actual header of the subspace block chain at a specific height
+    pub header: Header,
+    /// Global randomness after importing the header above.
+    /// This is same as the parent block unless update interval is met.
+    pub derived_global_randomness: Randomness,
+    /// Solution range after importing the header above.
+    /// This is same as the parent block unless update interval is met.
+    pub derived_solution_range: SolutionRange,
+    /// Salt after importing the header above.
+    /// This is same as the parent block unless update interval is met.
+    pub derived_salt: Salt,
+    /// Cumulative weight of chain until this header
+    pub total_weight: BlockWeight,
+}
+
+type HashOf<T> = <T as HeaderT>::Hash;
+
+/// Storage responsible for storing headers
+pub trait Storage<Header: HeaderT> {
+    /// Record size
+    fn record_size(&self) -> RecordSize;
+
+    /// Segment size
+    fn segment_size(&self) -> SegmentSize;
+
+    /// Queries a header at a specific block number or block hash
+    fn header(&self, query: HashOf<Header>) -> Option<HeaderExt<Header>>;
+
+    /// Stores the extended header.
+    /// as_best_header signifies of the header we are importing is considered best
+    fn store_header(&mut self, header_ext: HeaderExt<Header>, as_best_header: bool);
+
+    /// Returns the best known tip of the chain
+    fn best_header(&self) -> HeaderExt<Header>;
+}
+
+/// Error during the header import.
+#[derive(Debug, PartialEq, Eq)]
+pub enum ImportError<Hash> {
```

Comment:
Now it can make use of the new `ErrorDigestType` introduced in https://github.com/subspace/subspace/pull/694/files#diff-960199c19e15e765fcc8e0cbda704dfd1bac4ccc8f9b0f6383a37afac40aaf9cR151